lib/spack/spack/test/concretization/core.py

```diff
@@ -468,6 +468,41 @@ def test_mixing_compilers_only_affects_subdag(self):
             assert x.satisfies("%clang") is not expected_gcc
             assert x.satisfies("%gcc") is expected_gcc
 
+    def test_disable_mixing_prevents_mixing(self):
+        with spack.config.override("concretizer", {"compiler_mixing": False}):
+            with pytest.raises(spack.error.UnsatisfiableSpecError):
+                spack.concretize.concretize_one("dt-diamond%clang ^dt-diamond-bottom%gcc")
+
+    def test_disable_mixing_override_by_package(self):
+        with spack.config.override("concretizer", {"compiler_mixing": False}):
+            with spack.config.override(
+                "packages", {"dt-diamond-bottom": {"allow_compiler_mixing": True}}
+            ):
+                root = spack.concretize.concretize_one("dt-diamond%clang ^dt-diamond-bottom%gcc")
+                assert root.satisfies("%clang")
+                assert root["dt-diamond-bottom"].satisfies("%gcc")
+                assert root["dt-diamond-left"].satisfies("%clang")
+
+    def test_disable_mixing_reuse(self, mutable_database):
+        # Install a spec
+        left = spack.concretize.concretize_one("dt-diamond-left %gcc")
+        PackageInstaller([left.package], fake=True, explicit=True).install()
+        assert left.satisfies("%c=gcc")
+        lefthash = left.dag_hash()[:7]
+
+        # Now try to use it with compiler mixing disabled
+        with spack.config.override("concretizer", {"compiler_mixing": False}):
+            with pytest.raises(spack.error.UnsatisfiableSpecError):
+                spack.concretize.concretize_one(f"dt-diamond%clang ^/{lefthash}")
+
+    def test_disable_mixing_cfg_error(self):
+        # You can only set "allow_compiler_mixing" for individual
+        # packages, not the "all" pseudo package
+        with spack.config.override("concretizer", {"compiler_mixing": False}):
+            with pytest.raises(spack.config.ConfigFormatError):
+                with spack.config.override("packages", {"all": {"allow_compiler_mixing": True}}):
+                    pass
```

Comment:
```suggestion
                spack.config.set("packages", {"all": {"allow_compiler_mixing": True}})
```
No reason to create an extra code block for this